go-controller/pkg/ovn/address_set_sync.go

```diff
@@ -0,0 +1,348 @@
+	// these maps are used to track referenced that need to be replaced for every object type
+	aclsToUpdate := map[string]*nbdb.ACL{}
+	qosesToUpdate := map[string]*nbdb.QoS{}
+	lrpsToUpdate := map[string]*nbdb.LogicalRouterPolicy{}
+
+	for _, addrSetInfo := range addrSetsInfo {
+		if addrSetInfo.newAddrSet == nil {
+			// new address set wasn't built
+			if len(addrSetInfo.acls) == 0 && len(addrSetInfo.qoses) == 0 && len(addrSetInfo.lrps) == 0 {
+				// address set is stale and not referenced, clean up
+				ops, err = libovsdbops.DeleteAddressSetsOps(syncer.nbClient, ops, addrSetInfo.oldAddrSet)
+			} else {
+				syncer.ignoredAddressSets += 1
+			}
+			continue
+		}
+
+		oldName := addrSetInfo.oldAddrSet.ExternalIDs["name"]
+		// create updated address set
+		ops, err = libovsdbops.CreateOrUpdateAddressSetsOps(syncer.nbClient, ops, addrSetInfo.newAddrSet)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get update address set ops for address set %s: %v", oldName, err)
+		}
+		// delete old address set
+		ops, err = libovsdbops.DeleteAddressSetsOps(syncer.nbClient, ops, addrSetInfo.oldAddrSet)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get update address set ops for address set %s: %v", oldName, err)
+		}
+		oldHash := "$" + addrSetInfo.oldAddrSet.Name
+		newHash := "$" + addrSetInfo.newAddrSet.Name
+
+		for _, acl := range addrSetInfo.acls {
+			if _, ok := aclsToUpdate[acl.UUID]; !ok {
+				aclsToUpdate[acl.UUID] = acl
+			}
+			aclsToUpdate[acl.UUID].Match = strings.ReplaceAll(aclsToUpdate[acl.UUID].Match, oldHash, newHash)
+		}
+
+		for _, qos := range addrSetInfo.qoses {
+			if _, ok := qosesToUpdate[qos.UUID]; !ok {
+				qosesToUpdate[qos.UUID] = qos
+			}
+			qosesToUpdate[qos.UUID].Match = strings.ReplaceAll(qosesToUpdate[qos.UUID].Match, oldHash, newHash)
+		}
+
+		for _, lrp := range addrSetInfo.lrps {
+			if _, ok := lrpsToUpdate[lrp.UUID]; !ok {
+				lrpsToUpdate[lrp.UUID] = lrp
+			}
+			lrpsToUpdate[lrp.UUID].Match = strings.ReplaceAll(lrpsToUpdate[lrp.UUID].Match, oldHash, newHash)
+		}
+	}
+
+	for _, acl := range aclsToUpdate {
+		ops, err = libovsdbops.CreateOrUpdateACLsOps(syncer.nbClient, ops, acl)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get update acl ops: %v", err)
+		}
+	}
+	for _, qos := range qosesToUpdate {
+		ops, err = libovsdbops.CreateOrUpdateQoSesOps(syncer.nbClient, ops, qos)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get update qos ops: %v", err)
+		}
+	}
+	if len(lrpsToUpdate) > 0 {
+		// make sure lrps are referenced from the expected router
+		clusterRouter, err := libovsdbops.GetLogicalRouter(syncer.nbClient, &nbdb.LogicalRouter{Name: ovntypes.OVNClusterRouter})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get router %s from the db: %v", ovntypes.OVNClusterRouter, err)
+		}
+		clusterRouterPolicies := map[string]bool{}
+		for _, lrpUUID := range clusterRouter.Policies {
+			clusterRouterPolicies[lrpUUID] = true
+		}
```

Comment:
Hmm is there something I am missing here or would have it been a (probably better) alternative to have an update only method (instead of CreateOrUpdate) for the LRP.
I suggested if it was viable to do it in the predicate thinking there could probably be an easy way based on the LRP itself.

Reply:
So I ended up with `UpdateLogicalRouterPoliciesOps` method that has `ErrNotFound`=true. For ACL and QoS I will just use  existing methods